app/controllers/tokens_controller.rb

```diff
@@ -0,0 +1,69 @@
+class TokensController < Doorkeeper::TokensController
+  include ErrorResponses
+
+  def create
+    authenticate_with_facebook if signing_in_with_facebook?
+    authenticate_with_credentials if singing_in_regularly?
+  rescue Doorkeeper::Errors::DoorkeeperError,
+         Doorkeeper::Errors::InvalidGrantReuse,
+         Doorkeeper::OAuth::Error,
+         Koala::Facebook::AuthenticationError,
+         ActiveRecord::RecordNotFound => e
+    render_error e
+  end
+
+  private
+
+    def signing_in_with_facebook?
+      params[:username] == "facebook"
+    end
+
+    def singing_in_regularly?
+      !signing_in_with_facebook?
+    end
```

Comment:
I would consider making this more explicit. Isn't it `signing_in_with_email?` in this case?